### DISCUSSION TITLE
@pyscript/core PyScript Next into nuxt/vue app integration problems

### DISCUSSION THREAD

**Discussion by Manusan42** [pyscript/pyscript#1838] [Q&A] [answered]
Hello there,
trying to integrate the new experimental `@pyscript/core` module into a nuxt 3 (vuejs/typescript) project.

Sadly not a smooth experience so far. I got it somewhat working though.

So far tried with pretty much everything at default (so pyodide only):
- installed `@pyscript/core` with npm
- using within vue component 
`import * as pyscript from '@pyscript/core'` for triggering bootstrap for main thread or 
`import { PyWorker, hooks } from '@pyscript/core'` to use with `PyWorker` 
- `vue.copilerOptions.isCustomElement` in `nuxt.config.ts` to allow the `py-script`-tag
- `routeRules['/'].header` in `nuxt.config.ts` to allow workers

Had/have the following problems:

1. (kinda solved / nuxt/vite problem it seems) I could not change the vite build target for `nuxi dev` (config seems to only work with `nuxi build` at the moment) and was stuck on `es2020` which does not support top-level await which is `@pyscript/core` using apparently. Only thing which worked was using the `vite-plugin-top-level-await` module which transpiles the top-level await away.

2. (Probably nuxt-specific too) Using inline code within `<py-script>` gives always compiler error as the content is parsed as one line (stripped whitespace).  Only works if I wrap everything with `<pre>`.
As I prefer using a referenced python file, this is no big deal.

3. No matter what I tried, I got the automatic package installation not working within the main thread. Executing code within a `<py-script>` tag worked only without external package imports. I checked that the config is right and tried referencing it all possible ways. It did not work. It does not display the console log for loading the packages either (like it does with a PyWorker), but I do not know if that's normal.
But what works is the same config within a PyWorker, which I use for now.

4. Somethin is wrong with the provided Typescript-typings. 
![image](https://github.com/pyscript/pyscript/assets/12750571/3fbad1aa-e403-4b3a-ad56-24cb2cadb6f5)
It references `types/core` which exports only `any` stubs? But why? It seems all the types are there within `@pyscript/core/types/*.d.ts`-files. But you cannot access them? Is this for some reason by design or a bug? Or has it something to do with my setup?

5. I cannot use the hooks: It gives me `Cannot read properties of undefined (reading 'add')`. 
```typescript
<script setup lang="ts">
import { PyWorker, hooks } from '@pyscript/core';
hooks.onInterpreterReady.add((utils: any, element: any) => {
  console.debug(element, 'found', 'pyscript is ready');
});
</script>
```

6. So far I did not got hot-reload working with changes in referenced python files (probably need to configure the `chokidar`-watch), but for now I just reload the page. The thing is, it happens sporadically that after the worker started it gives a python error, that it cannot find the `pyscript` package, which is kinda strange. Sometimes it goes away after reloading the page again, and sometimes I need to restart the development server (`nuxi dev`).

7. More a general question: I know you can copy files to the worker with the config (which I did not try so far), but I did not find a way to access files created within the worker at worker runtime, that I can sync them back. The only idea I have, is to serialize the binary somehow and use the `sync` helper from `pyscript`. Is there another way?

That's it for now. 

PyWorker works, I also tested `pyscripts`'s `sync` and `when` which works too. Especially events to trigger something in the PyWorker which sends data back over sync which is assigned to a vue ref and used in template, is very nice.
Only thing is, the Worker takes a bit of time to start after every reload. Looking at the network tab, the libs seem to be fetched from cache, just the wasm load time is around 5 seconds for more complex libs like matplotlib, fonttools, numpy, pillow... which is probably normal.

I really like the project and hope there are some other people with some nuxt and/or @pyscript/core experience who could help me out a bit. 

Thanks

**Comment by Manusan42** [comment]
I dug a bit into pyodide and found a solution for point 7.
Using `from js import Blob, URL` and `from pyodide.ffi import to_js`, I can read the made file as binary data, get it's binary buffer, convert it to a js compatible buffer (int array) with `to_js` and construct a `Blob` from the web api accessed through the `js`-package. Then creating a blob-url which I send with sync to the main thread.

```python
file = open(file_path, "rb")
byte_buffer = file.read()
file.close()
blob = Blob.new([to_js(byte_buffer)], {type: "application/octet-stream"})
blob_url = URL.createObjectURL(blob)
sync.output({"blobUrl": blob_url})
```

On the javascript-end I can fetch the data which is probably a fetch from cache. Then use the binary data to provide a download or in my case i fetch it as text content:
```typescript
worker.sync.output = (data) => {
  (async () => {
    const blob = await fetch(data.blobUrl).then((r) => r.blob());
    out.value = await blob.text();
  })();
};
```

Seems a good approach, but I would still like to know if there is a better or more comfortable way.

**Comment by WebReflection** [comment, marked as answer]
lot of points but I can't really help much with your setup so point 1 and 2 will be skipped ... other points:

  3. the `<py-script>` both exists but it's mostly for backward compatibility and it warns if it contains some HTML tag. The recommended way to use PyScript is via `<script type="py">` which would solve everything you are witnessing unless your vite/setup complains also there about unrecognized type ... which I don't know, but "*this is the way*"
  4. there's literally no way to type the `xworker.sync` as that's al unknown until you define it and it's hard to explain to TS because it could be either sync (worker) or async (main) ... if anyone is willing to help out there, please file a PR. That being said, it's possible that the introduction of `stickyModule`, needed to avoid multiple bootstraps of the same module in case more projects re-bundle entirely PyScript and cause conflicts if landed in the same page, introduced a regression. I think I need to set `@type` explicitly so that's likely a bug we can fix.
  5. in latest projects hooks changed https://pyscript.github.io/polyscript/#hooks there is no `onInterpreterReady` anymore, it's eventually `main.onReady` and so goes for the `worker` counter part. Please read the updated docs.
  6. hot reload can be achieved by replacing the `<script type="py">` tag with an either shallow or deep clone of itself ... that would ~~re-bootstrap pyodide entirely~~ re-run the code as if it was a new script
  7. you can pass blobs as binary typed arrays from main to worker or vice-versa but the filesystem is not shared between main and worker. One day we could actually try to do that and orchestrate a filesystem that can share files and code across multiple interpreters too (i.e. MicroPython and Pyodide can share the same FS) but this is not planned (yet) and requires changes in both Pyodide and/or MicroPython WASM 
  8. yes, pyodide bootstrap takes 1 to 2 seconds, add external packages and it takes more. It doesn't really matter if it's a local build or not, it takes long time to load into the browser as WASM module ... unfortunately there's nothing we can do there

I hope some answer is helpful, feel free to ask more though.

**Reply by Manusan42** [reply]
Thank you for your quick reply.

So I ditched `@pyscript/core` and installed `polyscript` instead.

Sadly there is no easy way to get script tags into vue-components, as the template is just not meant for it. There are workarounds of course, but I decided to go with workers only as I need to do some heavy computation stuff anyway.

I then inspected the documentation at https://pyscript.github.io/polyscript/ and figured most stuff out.

Surely more of the typings can be provided for the imported stuff. For the sync-methods I wrote my own interface which extends from `Worker`. I think this makes sense anyway as I'm the one who is creating the custom functions. 

I have a working XWorker which can communicate bidirectional with the sync-interface. I use then blob-urls to access binary data created form the worker.

https://pyscript.github.io/polyscript/#just-the-xworker says I only need `import { Hook, XWorker } from 'polyscript/xworker'`,
but I cant figure out how to get the Hook working.
I just tried to instanciate it like this:
```typescript
new Hook('pyodide', {
  worker: {
    onReady: () => {
      console.log('pyodide is ready');
    },
  },
});
```
This seems not to work. Do I need to add the hook-instance somewhere?

**Reply by WebReflection** [reply]
`Hook` is really an internal thing for advanced usage but you can see in polyscript how that's used or even in PyScript ... basically you need to pass `null` and the list of hooks but that should be an `XWorker.call(hookInstance, { options })` which you can return out of a wrapper function or an `XWorker` extended class within its constructor.

**Reply by WebReflection** [reply]
FYI types should be back in once this gets approved: https://github.com/pyscript/pyscript/pull/1843

**Reply by Manusan42** [reply]
Thank you very much, I've got everything working to my satisfaction for now.

**Comment by bugzpodder** [comment]
> 7\. More a general question: I know you can copy files to the worker with the config (which I did not try so far), but I did not find a way to access files created within the worker at worker runtime, that I can sync them back. The only idea I have, is to serialize the binary somehow and use the `sync` helper from `pyscript`. Is there another way?

yeah sounds about right. 

> Sadly there is no easy way to get script tags into vue-components, as the template is just not meant for it. There are workarounds of course, but I decided to go with workers only as I need to do some heavy computation stuff anyway.

 I had to go through the same exercise as you and decided to maintain a library on top of polyscript.  feel free to look at it for inspiration: https://bugzpodder.github.io/pysandbox/

**Reply by Manusan42** [reply]
Thanks. For now I have what I need, but I might take a look at some point, seems also a nice project you've made there.
